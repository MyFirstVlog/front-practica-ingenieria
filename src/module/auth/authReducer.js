import { types } from "../../constants/types";

export const authReducer = (state, action) => {
	switch (action.type) {
		case types.login:
			return {
				...state,
				logged: true,
				email: action.payload.email,
			};

		case types.logout:
			return {
				...state,
				logged: false,
				email: null,
			};

		default:
			return state;
	}
};
