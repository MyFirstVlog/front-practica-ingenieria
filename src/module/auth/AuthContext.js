import React from "react";
import { createContext, useReducer } from "react";
import { authReducer } from "./authReducer";

export const AuthContext = createContext();

const initialState = {
	logged: false,
	email: "",
};

export const AuthProvider = ({ children }) => {
	const [authState, dispatch] = useReducer(authReducer, initialState);

	return (
		<AuthContext.Provider
			value={{
				authState,
				dispatch,
			}}
		>
			{children}
		</AuthContext.Provider>
	);
};
