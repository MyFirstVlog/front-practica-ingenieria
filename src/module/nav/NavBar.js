import React, { useContext, useState } from "react";
import { useHistory } from "react-router-dom";
import { NavLink } from "react-router-dom";
import { Link } from "react-router-dom";
import { types } from "../../constants/types";
import { AuthContext } from "../auth/AuthContext";

export const NavBar = () => {
	const [isNavCollapsed, setIsNavCollapsed] = useState(true);

	const handleNavCollapse = () => setIsNavCollapsed(!isNavCollapsed);

	const history = useHistory();

	const { dispatch } = useContext(AuthContext);

	const handleLogout = () => {
		dispatch({
			type: types.logout,
		});

		history.replace("/login");
	};

	return (
		<nav
			className="navbar navbar-expand-lg navbar-light"
			style={{ backgroundColor: "#e3f2fd" }}
		>
			<div className="container-fluid">
				<Link className="navbar-brand" to="/">
					Gestionappte
				</Link>
				<button
					className="navbar-toggler"
					type="button"
					data-bs-toggle="collapse"
					data-bs-target="#navbarColor03"
					aria-controls="navbarColor03"
					aria-expanded={!isNavCollapsed ? true : false}
					aria-label="Toggle navigation"
					onClick={handleNavCollapse}
				>
					<span className="navbar-toggler-icon"></span>
				</button>
				<div
					className={`${isNavCollapsed ? "collapse" : ""}  navbar-collapse`}
					id="navbarColor03"
				>
					<ul className="navbar-nav me-auto mb-2 mb-lg-0">
						<li className="nav-item">
							<NavLink
								activeClassName="active"
								className="nav-item nav-link"
								exact
								to="/scheduling"
							>
								Agendamiento
							</NavLink>
						</li>
						<li className="nav-item">
							<NavLink
								activeClassName="active"
								className="nav-item nav-link"
								exact
								to="/bookings"
							>
								Reservaciones
							</NavLink>
						</li>
						<li className="nav-item">
							<NavLink
								activeClassName="active"
								className="nav-item nav-link"
								exact
								to="/stats"
							>
								Estadisticas
							</NavLink>
						</li>
					</ul>
					<div className="d-flex">
						<ul className="navbar-nav mr-auto">
							<span className="nav-item nav-link text-info">Alejandro</span>

							<button className="nav-item nav-link btn" onClick={handleLogout}>
								Logout
							</button>
						</ul>
					</div>
				</div>
			</div>

			{/* poner condicional  abajo para que solo aparezca antes del button y arriba despyues de activar el breakpoint */}
		</nav>
	);
};
