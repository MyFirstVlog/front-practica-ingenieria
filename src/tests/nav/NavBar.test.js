import React from "react";
import { mount } from "enzyme";
import { MemoryRouter, Router } from "react-router-dom";
import "@testing-library/jest-dom";

import { AuthContext } from "../../module/auth/AuthContext";
import { NavBar } from "../../module/nav/NavBar";

describe("should test NavBar component", () => {
	const historyMock = {
		push: jest.fn(),
		replace: jest.fn(),
		location: {},
		listen: jest.fn(),
		createHref: jest.fn(),
	};

	const contextValue = {
		dispatch: jest.fn(),
		authState: {
			logged: true,
			email: "example1@example.com",
		},
	};
	const wrapper = mount(
		<AuthContext.Provider value={contextValue}>
			<MemoryRouter>
				<Router history={historyMock}>
					<NavBar />
				</Router>
			</MemoryRouter>
		</AuthContext.Provider>
	);

	afterEach(() => {
		jest.clearAllMocks();
	});
	test("should test logout", () => {
		wrapper.find("button").at(1).simulate("click");
		expect(contextValue.dispatch).toHaveBeenCalledWith({
			type: "[logout] Logout",
		});
	});
	test("should test collapse navbar", () => {
		Object.defineProperty(window, "innerHeight", {
			writable: true,
			configurable: true,
			value: 150,
		});

		window.dispatchEvent(new Event("resize"));

		expect(window.innerHeight).toBe(150);

		wrapper.find("button").at(0).simulate("click");
		expect(wrapper).toMatchSnapshot();
	});
});
