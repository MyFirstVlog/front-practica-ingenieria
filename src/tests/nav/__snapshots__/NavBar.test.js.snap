// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should test NavBar component should test collapse navbar 1`] = `
<nav
  className="navbar navbar-expand-lg navbar-light"
  style={
    Object {
      "backgroundColor": "#e3f2fd",
    }
  }
>
  <div
    className="container-fluid"
  >
    <a
      className="navbar-brand"
      onClick={[Function]}
    >
      Gestionappte
    </a>
    <button
      aria-controls="navbarColor03"
      aria-expanded={true}
      aria-label="Toggle navigation"
      className="navbar-toggler"
      data-bs-target="#navbarColor03"
      data-bs-toggle="collapse"
      onClick={[Function]}
      type="button"
    >
      <span
        className="navbar-toggler-icon"
      />
    </button>
    <div
      className="  navbar-collapse"
      id="navbarColor03"
    >
      <ul
        className="navbar-nav me-auto mb-2 mb-lg-0"
      >
        <li
          className="nav-item"
        >
          <a
            aria-current={null}
            className="nav-item nav-link"
            onClick={[Function]}
          >
            Agendamiento
          </a>
        </li>
        <li
          className="nav-item"
        >
          <a
            aria-current={null}
            className="nav-item nav-link"
            onClick={[Function]}
          >
            Reservaciones
          </a>
        </li>
        <li
          className="nav-item"
        >
          <a
            aria-current={null}
            className="nav-item nav-link"
            onClick={[Function]}
          >
            Estadisticas
          </a>
        </li>
      </ul>
      <div
        className="d-flex"
      >
        <ul
          className="navbar-nav mr-auto"
        >
          <span
            className="nav-item nav-link text-info"
          >
            Alejandro
          </span>
          <button
            className="nav-item nav-link btn"
            onClick={[Function]}
          >
            Logout
          </button>
        </ul>
      </div>
    </div>
  </div>
</nav>
`;
